flink-runtime/src/main/java/org/apache/flink/runtime/state/LocalStateRegistry.java

```diff
@@ -0,0 +1,117 @@
+    /**
+     * All registered handles. (handle,checkpointID) represents a handle and the latest checkpoint
+     * that refer to this handle.
+     */
+    private final Map<StreamStateHandle, Long> registeredHandles;
+
+    /** This flag indicates whether the registry is open or if close() was called. */
+    private boolean open;
+
+    /** Executor for async state deletion */
+    private final Executor asyncDisposalExecutor;
+
+    public LocalStateRegistry(Executor ioExecutor) {
+        this.registeredHandles = new HashMap<>();
+        this.asyncDisposalExecutor = ioExecutor;
+        this.open = true;
+    }
+
+    public StreamStateHandle register(StreamStateHandle handle, long checkpointID) {
+        synchronized (registeredHandles) {
+            checkState(open, "Attempt to register state to closed LocalStateRegistry.");
+            if (registeredHandles.containsKey(handle)) {
+                long pre = registeredHandles.get(handle);
+                if (checkpointID > pre) {
+                    registeredHandles.put(handle, checkpointID);
+                }
+            } else {
+                registeredHandles.put(handle, checkpointID);
+            }
+        }
+        return handle;
+    }
+
+    public void unRegister(long upTo) {
+        List<StreamStateHandle> handles = new ArrayList<>();
+        synchronized (registeredHandles) {
+            Iterator<Entry<StreamStateHandle, Long>> iterator =
+                    registeredHandles.entrySet().iterator();
+            while (iterator.hasNext()) {
+                Entry<StreamStateHandle, Long> entry = iterator.next();
+                if (entry.getValue() < upTo) {
+                    handles.add(entry.getKey());
+                    iterator.remove();
+                }
+            }
+        }
+        for (StreamStateHandle handle : handles) {
+            scheduleAsyncDelete(handle);
+        }
+    }
+
+    private void scheduleAsyncDelete(StreamStateHandle streamStateHandle) {
+        if (streamStateHandle != null) {
+            LOG.trace("Scheduled delete of state handle {}.", streamStateHandle);
+            Runnable discardRunner =
+                    () -> {
+                        try {
+                            streamStateHandle.discardState();
+                        } catch (Exception exception) {
+                            LOG.warn(
+                                    "A problem occurred during asynchronous disposal of a stream handle {}.",
+                                    streamStateHandle);
+                        }
+                    };
+            try {
+                asyncDisposalExecutor.execute(discardRunner);
+            } catch (RejectedExecutionException ex) {
+                discardRunner.run();
+            }
+        }
+    }
+
+    @Override
+    public void close() {
+        synchronized (registeredHandles) {
+            open = false;
+        }
+    }
```

Comment:
As mentioned above, this method should probably discard the remaining state.

If not, and if this class will stay per task (likely not); I'd remove `open` because it requires adding `syncrhonized` sections, which IMO doesn't worth enforcing the lifecycle without any cleanup.

Reply:
Yes, it doesn't worth enforcing the lifecycle without any cleanup, removed it.